Improve documentation on saga behaviour when using Sql Server transport and persistence (#6690)
Particular/docs.particular.net: persistence/sql/sqlserver-combining-persistence-wth-transport.md

---
title: Combining MS SQL Server Persistence and Transport
component: SqlPersistence
reviewed: 2024-06-10
related:
  - persistence/sql/dialect-mysql
  - transports/sql
---

## Connection behavior

When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.

### Without Outbox

SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
:-:|:-:|:-:|:-:

> TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db

Reviewer: Note, I'm not sure if &lt;sup&gt; will work on docs site

```suggestion
TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
```

Author: @abparticular it seems to work ok - I tried it locally first then pushed the change.

> AtomicSendsWithReceive |  ✅| SQLP uses isolated transaction for send and receive | Persistence Db

Reviewer: ```suggestion
AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

> ReceiveOnly |  ✅| SQLP uses isolated transaction for receive | Persistence Db

Reviewer: ```suggestion
ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
```

> None |  ✅| No transactions | Persistence Db

Reviewer: ```suggestion
None |  ✅| No transactions | Persistence DB

<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```

#### Explicitly opting out of connection sharing when not using the Outbox

When an endpoint uses SQL Persistence combined with the SQL Server Transport without the [Outbox](/nservicebus/outbox/), the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming of the message that triggered it.

partial: Connection

### With Outbox

SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
:-:|:-:|:-:|:-:
TransactionScope |  ✅|  Not supported | N/A
AtomicSendsWithReceive |  ✅| Not supported | N/A

> ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence Db

Reviewer: ```suggestion
ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence DB
```

---
title: Combining MS SQL Server Persistence and Transport
component: SqlPersistence
reviewed: 2024-06-10
related:
  - persistence/sql/dialect-mysql
  - transports/sql
---

## Connection behavior

> When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.

Reviewer 2: ```suggestion
When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is enabled or disabled. This influences where the saga data is stored.
```

### Without Outbox

SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
:-:|:-:|:-:|:-:

> TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>

Reviewer 2: What is `SQLT`, according to google https://www.google.com/search?q=sqlt
I guess my point is that the acronym, is known for something else, I think we need to change it

Author: I have changed it to SQL Transport

> AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB

Reviewer: ```suggestion
SendsAtomicWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
None |  ✅| No transactions | Persistence DB

> <sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`

Reviewer 2: Would this only work on Windows OSs?

Reviewer 2: and if so, should we point it out

> <sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`

Reviewer 2: ```suggestion
<sup>1</sup> - Requires .NET Framework, or .NET 8 [`ImplicitDistributedTransactions`](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.transactionmanager.implicitdistributedtransactions?view=net-8.0) set to `true`.
```

#### Explicitly opting out of connection sharing when not using the Outbox

> When an endpoint uses SQL Persistence combined with the SQL Server Transport without the [Outbox](/nservicebus/outbox/), the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming of the message that triggered it.

Reviewer 2: ```suggestion
When an endpoint uses SQL Persistence combined with the SQL Server Transport with [Outbox](/nservicebus/outbox/) disabled, the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming the message that triggered it.
```

partial: Connection

### With Outbox

SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
:-:|:-:|:-:|:-:

> TransactionScope |  ✅|  Not supported | N/A

Reviewer 2: I don't see the purpose of ✅ in the table?

Author: I have removed them.
